In [ ]:
!pip install gplearn

In [6]:
from gplearn.genetic import SymbolicTransformer
from sklearn.utils import check_random_state
from sklearn.datasets import load_boston
import numpy as np
import pandas as pd
from sklearn.linear_model import Ridge

In [ ]:
rng = check_random_state(0)
boston = load_boston()
perm = rng.permutation(boston.target.size)
boston.data = boston.data[perm]
boston.target = boston.target[perm]

In [8]:
est = Ridge()
est.fit(boston.data[:300, :], boston.target[:300])
print(est.score(boston.data[300:, :], boston.target[300:]))

0.759319453049884


In [9]:
function_set = ['add', 'sub', 'mul', 'div',
                'sqrt', 'log', 'abs', 'neg', 'inv',
                'max', 'min']
gp = SymbolicTransformer(generations=20, population_size=2000,
                         hall_of_fame=100, n_components=10,
                         function_set=function_set,
                         parsimony_coefficient=0.0005,
                         max_samples=0.9, verbose=1,
                         random_state=0, n_jobs=3)
gp.fit(boston.data[:300, :], boston.target[:300])

    |   Population Average    |             Best Individual              |
---- ------------------------- ------------------------------------------ ----------
 Gen   Length          Fitness   Length          Fitness      OOB Fitness  Time Left
   0    11.04         0.339876        6         0.822502         0.675124      1.84m
   1     6.91         0.593562        7         0.836993         0.602468      1.23m
   2     5.07         0.730093        8          0.84063         0.704017      2.11m
   3     5.22         0.735525        5         0.847019         0.628351      1.29m
   4     6.24         0.734679       10         0.856612         0.565138      1.43m
   5     8.23         0.721433       18          0.85677         0.728095      1.15m
   6    10.20         0.717937       14         0.875233         0.619693      1.17m
   7    11.84         0.720667       14         0.875927         0.609363      1.02m
   8    12.56         0.733019       27         0.881705         0.390121  

SymbolicTransformer(function_set=['add', 'sub', 'mul', 'div', 'sqrt', 'log',
                                  'abs', 'neg', 'inv', 'max', 'min'],
                    max_samples=0.9, n_jobs=3, parsimony_coefficient=0.0005,
                    population_size=2000, random_state=0, verbose=1)

In [10]:
gp_features = gp.transform(boston.data)
new_boston = np.hstack((boston.data, gp_features))

In [11]:
est = Ridge()
est.fit(new_boston[:300, :], boston.target[:300])
print(est.score(new_boston[300:, :], boston.target[300:]))

0.8418372105182054


We see that when we applied Ridge Estimator to previous dataset without any symbolic transformer , The R-square value was lower i.e approx 75 % of variance , however when we applied the Symbolic transformer to the new dataset we get the R-square value as 0.84 i.e approx 84 % of variance of data in the estimation